### ISSUE TITLE
[New rule] Forbidden types on data classes

### ISSUE THREAD

**Issue by BraisGabin** [detekt/detekt#5521]
## Expected Behavior of the rule
non compilatn:
```kotlin
data class Foo(
  val throwable: Throwable,
)
```
```kotlin
data class Foo(
  val painter: () -> Painter,
)
```

## Context
There are some types of properties that should not be allowed inside a `data class`. Two examples are `Throwable`s and lambdas. Those classes doesn't implement the `equals` so it have little sense to add them inside a `data class`.

In general I see that lots of devs use `data class` "by default" without thinking about the implications.

For me the classes that I defined could have total sense but the `data` modifier should be removed.

I think that the rule should also allow to define other type of properties to make it more flexible.

**Comment by hfhbd** [comment]
What about arrays? You always get a warning too

**Comment by eygraber** [comment]
For `Throwable` shouldn't that be up to the implementation? i.e. if I have a Throwable subclass that does implement `equals`:

```
data class EqualsThrowable(val m: String) : Throwable(m)
```

then there's nothing wrong with:

```
data class Foo(throwable: Throwable)

val t1 = EqualsThrowable("Hello")
val t2 = EqualsThrowable("Hello")

Foo(t1) == Foo(t2)
```

**Comment by eygraber** [comment]
The point I was trying to make was that it's possible to have a `Throwable` that does implement `equals` correctly (using a data class for it being a separately bad idea).

If you have a common ancestor below `Throwable` then you could use that as the type in the data class. But there could be a scenario where there is no common ancestor other than `Throwable`.

**Comment by BraisGabin** [comment]
I'm not sure if I like a `data class` that extends `Throwable` ([more about that here](https://kotlinlang.slack.com/archives/C88E12QH4/p1669019792378529)).

The idea behind of this rule is "all the properties of a `data class` should implement `equals`". The problem is that we can't ensure that with static analysis. An example would be an `interface`. We can't know if the `subtype` of that interface implements `equals`.

So the best we can do is to create a rule were the user can add classes that s/he knows that doesn't implement `equals` and just check for those.

Other improvement would be to check if the value that we receive is `final` or `sealed`. If that's the case I assume that we could ensure that.

**Comment by BraisGabin** [comment]
I'm going to keep this closed. It's clearly a controversial rule. I'll probably create it as a custom rule but it doesn't seem to fit inside the default rule set.
